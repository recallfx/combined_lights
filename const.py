"""Constants for the Combined lights integration."""

DOMAIN = "combined_lights"

# Platforms to be set up.
NUMBER_PLATFORM = "number"

# Configuration keys used in the config flow.
CONF_NAME = "name"
CONF_BACKGROUND_LIGHTS = "background_lights"
CONF_FEATURE_LIGHTS = "feature_lights"
CONF_CEILING_LIGHTS = "ceiling_lights"

# Advanced configuration keys for breakpoints and brightness ranges
CONF_BREAKPOINTS = "breakpoints"  # [30, 60, 90] - slider positions where zones activate
CONF_BACKGROUND_BRIGHTNESS_RANGES = (
    "background_brightness_ranges"  # [[1,50], [51,70], [71,80], [81,100]]
)
CONF_FEATURE_BRIGHTNESS_RANGES = (
    "feature_brightness_ranges"  # [[1,50], [51,70], [71,100]]
)
CONF_CEILING_BRIGHTNESS_RANGES = "ceiling_brightness_ranges"  # [[1,50], [51,100]]

# Default configuration matching your original request
DEFAULT_BREAKPOINTS = [30, 60, 90]  # 1-30%, 31-60%, 61-90%, 91-100%

# Default brightness ranges for each zone at each stage
DEFAULT_BACKGROUND_BRIGHTNESS_RANGES = [
    [1, 50],  # Stage 1 (1-30%): background at 1-50%
    [51, 70],  # Stage 2 (31-60%): background at 51-70%
    [71, 80],  # Stage 3 (61-90%): background at 71-80%
    [81, 100],  # Stage 4 (91-100%): background at 81-100%
]

DEFAULT_FEATURE_BRIGHTNESS_RANGES = [
    [0, 0],  # Stage 1: feature lights off
    [1, 50],  # Stage 2 (31-60%): feature at 1-50%
    [51, 70],  # Stage 3 (61-90%): feature at 51-70%
    [71, 100],  # Stage 4 (91-100%): feature at 71-100%
]

DEFAULT_CEILING_BRIGHTNESS_RANGES = [
    [0, 0],  # Stage 1: ceiling lights off
    [0, 0],  # Stage 2: ceiling lights off
    [1, 50],  # Stage 3 (61-90%): ceiling at 1-50%
    [51, 100],  # Stage 4 (91-100%): ceiling at 51-100%
]
