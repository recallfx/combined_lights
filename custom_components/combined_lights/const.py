"""Constants for the Combined lights integration."""

DOMAIN = "combined_lights"

# Platforms to be set up.
NUMBER_PLATFORM = "number"

# Configuration keys used in the config flow.
CONF_NAME = "name"

# Configuration keys updated to use stage-based terminology
CONF_STAGE_1_LIGHTS = "stage_1_lights"
CONF_STAGE_2_LIGHTS = "stage_2_lights"
CONF_STAGE_3_LIGHTS = "stage_3_lights"
CONF_STAGE_4_LIGHTS = "stage_4_lights"

# Advanced configuration keys for breakpoints and brightness ranges
CONF_BREAKPOINTS = "breakpoints"  # [25, 50, 75] - slider positions where zones activate
CONF_STAGE_1_BRIGHTNESS_RANGES = (
    "stage_1_brightness_ranges"  # [[1,40], [41,60], [61,80], [81,100]]
)
CONF_STAGE_2_BRIGHTNESS_RANGES = (
    "stage_2_brightness_ranges"  # [[0,0], [1,40], [41,60], [61,100]]
)
CONF_STAGE_3_BRIGHTNESS_RANGES = (
    "stage_3_brightness_ranges"  # [[0,0], [0,0], [1,50], [51,100]]
)
CONF_STAGE_4_BRIGHTNESS_RANGES = (
    "stage_4_brightness_ranges"  # [[0,0], [0,0], [0,0], [1,100]]
)

# Brightness curve configuration
CONF_BRIGHTNESS_CURVE = "brightness_curve"  # "linear", "quadratic", or "cubic"

# Brightness curve types
CURVE_LINEAR = "linear"
CURVE_QUADRATIC = "quadratic"  # More precision at low brightness
CURVE_CUBIC = "cubic"  # Even more precision at low brightness

# Default curve - linear provides most predictable behavior
DEFAULT_BRIGHTNESS_CURVE = CURVE_LINEAR

# Default configuration matching your original request
DEFAULT_BREAKPOINTS = [25, 50, 75]  # 1-25%, 26-50%, 51-75%, 76-100%

# Default brightness ranges for each zone at each stage (4 stages now instead of 4)
DEFAULT_STAGE_1_BRIGHTNESS_RANGES = [
    [1, 40],  # Stage 1 (1-25%): lights at 1-40%
    [41, 60],  # Stage 2 (26-50%): lights at 41-60%
    [61, 80],  # Stage 3 (51-75%): lights at 61-80%
    [81, 100],  # Stage 4 (76-100%): lights at 81-100%
]

DEFAULT_STAGE_2_BRIGHTNESS_RANGES = [
    [0, 0],  # Stage 1: lights off
    [1, 40],  # Stage 2 (26-50%): lights at 1-40%
    [41, 60],  # Stage 3 (51-75%): lights at 41-60%
    [61, 100],  # Stage 4 (76-100%): lights at 61-100%
]

DEFAULT_STAGE_3_BRIGHTNESS_RANGES = [
    [0, 0],  # Stage 1: lights off
    [0, 0],  # Stage 2: lights off
    [1, 50],  # Stage 3 (51-75%): lights at 1-50%
    [51, 100],  # Stage 4 (76-100%): lights at 51-100%
]

DEFAULT_STAGE_4_BRIGHTNESS_RANGES = [
    [0, 0],  # Stage 1: lights off
    [0, 0],  # Stage 2: lights off
    [0, 0],  # Stage 3: lights off
    [1, 100],  # Stage 4 (76-100%): lights at 1-100%
]
