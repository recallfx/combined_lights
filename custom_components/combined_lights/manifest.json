{
  "domain": "combined_lights",
  "name": "Combined Lights",
  "after_dependencies": ["light"],
  "codeowners": ["@recallfx"],
  "config_flow": true,
  "dependencies": [],
  "documentation": "https://github.com/recallfx/combined_lights",
  "integration_type": "service",
  "iot_class": "local_push",
  "issue_tracker": "https://github.com/recallfx/combined_lights/issues",
  "version": "2.2.0"
}