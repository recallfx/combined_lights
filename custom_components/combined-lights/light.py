"""Light platform for Combined Lights integration."""

from __future__ import annotations

import logging
from typing import Any

from homeassistant.components.light import ATTR_BRIGHTNESS, ColorMode, LightEntity
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import EVENT_STATE_CHANGED
from homeassistant.core import Event, HomeAssistant, callback
from homeassistant.exceptions import ServiceNotFound
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import (
    CONF_BACKGROUND_BRIGHTNESS_RANGES,
    CONF_BACKGROUND_LIGHTS,
    CONF_BREAKPOINTS,
    CONF_BRIGHTNESS_CURVE,
    CONF_CEILING_BRIGHTNESS_RANGES,
    CONF_CEILING_LIGHTS,
    CONF_FEATURE_BRIGHTNESS_RANGES,
    CONF_FEATURE_LIGHTS,
    CURVE_CUBIC,
    CURVE_QUADRATIC,
    DEFAULT_BACKGROUND_BRIGHTNESS_RANGES,
    DEFAULT_BREAKPOINTS,
    DEFAULT_BRIGHTNESS_CURVE,
    DEFAULT_CEILING_BRIGHTNESS_RANGES,
    DEFAULT_FEATURE_BRIGHTNESS_RANGES,
)

_LOGGER = logging.getLogger(__name__)


# Utility functions to eliminate code duplication
def get_config_value(entry: ConfigEntry, key: str, default: Any) -> Any:
    """Get configuration value with default fallback."""
    return entry.data.get(key, default)


def get_light_zones(entry: ConfigEntry) -> dict[str, list[str]]:
    """Get all light zones from configuration."""
    return {
        "background": get_config_value(entry, CONF_BACKGROUND_LIGHTS, []),
        "feature": get_config_value(entry, CONF_FEATURE_LIGHTS, []),
        "ceiling": get_config_value(entry, CONF_CEILING_LIGHTS, []),
    }


def get_brightness_ranges(entry: ConfigEntry) -> dict[str, list[list[int]]]:
    """Get all brightness ranges from configuration."""
    return {
        "background": get_config_value(
            entry,
            CONF_BACKGROUND_BRIGHTNESS_RANGES,
            DEFAULT_BACKGROUND_BRIGHTNESS_RANGES,
        ),
        "feature": get_config_value(
            entry, CONF_FEATURE_BRIGHTNESS_RANGES, DEFAULT_FEATURE_BRIGHTNESS_RANGES
        ),
        "ceiling": get_config_value(
            entry, CONF_CEILING_BRIGHTNESS_RANGES, DEFAULT_CEILING_BRIGHTNESS_RANGES
        ),
    }


async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up Combined Lights light entity."""
    # Create a single combined light entity
    async_add_entities([CombinedLight(entry)], True)


class CombinedLight(LightEntity):
    """Combined Light entity that controls multiple light zones."""

    def __init__(self, entry: ConfigEntry) -> None:
        """Initialize the Combined Light."""
        self._entry = entry
        self._attr_name = get_config_value(entry, "name", "Combined Lights")
        self._attr_unique_id = f"{entry.entry_id}_combined_light"
        self._attr_is_on = False
        self._attr_brightness = 255
        self._attr_supported_color_modes = {ColorMode.BRIGHTNESS}
        self._attr_color_mode = ColorMode.BRIGHTNESS
        self._remove_listener = None
        self._target_brightness = 255  # Track the intended brightness

        # No device info - entity will be created without a device

    async def async_added_to_hass(self) -> None:
        """Entity added to Home Assistant."""
        await super().async_added_to_hass()

        # Initialize target brightness based on current state
        if self.is_on and not hasattr(self, "_target_brightness_initialized"):
            # If lights are already on, estimate target brightness from current state
            self._sync_target_brightness_from_lights()
            self._target_brightness_initialized = True

        # Get all controlled lights
        all_lights = self._get_all_controlled_lights()

        # Listen for state changes of controlled lights
        @callback
        def light_state_changed(event: Event) -> None:
            """Handle controlled light state changes."""
            entity_id = event.data.get("entity_id")
            if entity_id in all_lights:
                # Only schedule update, don't recalculate target brightness
                # This prevents feedback loops while still updating the is_on state
                self.async_schedule_update_ha_state()

        self._remove_listener = self.hass.bus.async_listen(
            EVENT_STATE_CHANGED, light_state_changed
        )

    def _sync_target_brightness_from_lights(self) -> None:
        """Sync target brightness from actual light states (used during initialization)."""
        # Get configuration
        breakpoints = get_config_value(
            self._entry, CONF_BREAKPOINTS, DEFAULT_BREAKPOINTS
        )
        light_zones = get_light_zones(self._entry)

        # Get average brightness from each zone
        zone_brightness = {
            zone: self._get_average_brightness(lights)
            for zone, lights in light_zones.items()
        }

        # Simple heuristic: use the highest brightness from active zones
        # and map it to an appropriate target brightness
        max_brightness = 0
        if zone_brightness["ceiling"]:
            # Ceiling lights suggest we're in stage 3 or 4
            max_brightness = max(
                max_brightness,
                breakpoints[2] * 255 // 100 + zone_brightness["ceiling"] // 2,
            )
        elif zone_brightness["feature"]:
            # Feature lights suggest we're in stage 2 or 3
            max_brightness = max(
                max_brightness,
                breakpoints[1] * 255 // 100 + zone_brightness["feature"] // 2,
            )
        elif zone_brightness["background"]:
            # Background lights suggest we're in stage 1 or 2
            max_brightness = max(max_brightness, zone_brightness["background"])

        if max_brightness > 0:
            self._target_brightness = min(255, max_brightness)
            self._attr_brightness = self._target_brightness

    async def async_will_remove_from_hass(self) -> None:
        """Entity removed from Home Assistant."""
        if self._remove_listener:
            self._remove_listener()
        await super().async_will_remove_from_hass()

    def _get_all_controlled_lights(self) -> list[str]:
        """Get all controlled light entity IDs."""
        light_zones = get_light_zones(self._entry)
        return sum(light_zones.values(), [])  # Flatten all zone lists

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return True

    @property
    def is_on(self) -> bool:
        """Return true if any controlled light is on."""
        all_lights = self._get_all_controlled_lights()
        for entity_id in all_lights:
            state = self.hass.states.get(entity_id)
            if state and state.state == "on":
                return True
        return False

    @property
    def brightness(self) -> int | None:
        """Return the target brightness of the combined light."""
        if not self.is_on:
            return None
        return self._target_brightness

    def _get_average_brightness(self, light_entities: list[str]) -> int | None:
        """Get average brightness of lights that are on."""
        brightness_values = []
        for entity_id in light_entities:
            state = self.hass.states.get(entity_id)
            if state and state.state == "on":
                brightness = state.attributes.get("brightness")
                if brightness is not None:
                    brightness_values.append(brightness)

        return (
            int(sum(brightness_values) / len(brightness_values))
            if brightness_values
            else None
        )

    async def async_turn_on(self, **kwargs: Any) -> None:
        """Turn on the combined light."""
        if ATTR_BRIGHTNESS in kwargs:
            self._target_brightness = kwargs[ATTR_BRIGHTNESS]
            self._attr_brightness = kwargs[ATTR_BRIGHTNESS]

        self._attr_is_on = True

        # Get configuration using utility functions
        breakpoints = get_config_value(
            self._entry, CONF_BREAKPOINTS, DEFAULT_BREAKPOINTS
        )
        light_zones = get_light_zones(self._entry)
        brightness_ranges = get_brightness_ranges(self._entry)

        # Convert brightness to percentage (0-100)
        brightness_pct = (self._target_brightness / 255.0) * 100

        # Determine which stage we're in based on breakpoints
        stage = self._get_stage_from_brightness(brightness_pct, breakpoints)

        # Calculate and apply brightness for each zone
        zone_brightness = {}
        for zone_name in ["background", "feature", "ceiling"]:
            zone_brightness[zone_name] = self._calculate_zone_brightness_from_config(
                brightness_pct, stage, brightness_ranges[zone_name], breakpoints
            )

        # Control all zones
        await self._control_all_zones(light_zones, zone_brightness)

        _LOGGER.info(
            "Combined light turned on - overall: %s%% (stage %s) | background: %s%% | feature: %s%% | ceiling: %s%%",
            int(brightness_pct),
            stage + 1,
            int(zone_brightness["background"])
            if zone_brightness["background"] > 0
            else 0,
            int(zone_brightness["feature"]) if zone_brightness["feature"] > 0 else 0,
            int(zone_brightness["ceiling"]) if zone_brightness["ceiling"] > 0 else 0,
        )

        self.async_write_ha_state()

    async def _control_all_zones(
        self, light_zones: dict[str, list[str]], zone_brightness: dict[str, float]
    ) -> None:
        """Control all light zones based on calculated brightness values."""
        for zone_name, lights in light_zones.items():
            if not lights:
                continue

            brightness = zone_brightness[zone_name]
            if brightness > 0:
                await self._control_lights(lights, brightness / 100.0)
            else:
                await self._turn_off_lights(lights)

    def _get_stage_from_brightness(
        self, brightness_pct: float, breakpoints: list[int]
    ) -> int:
        """Determine stage based on brightness percentage and breakpoints."""
        if brightness_pct <= breakpoints[0]:
            return 0  # Stage 1: 0% to breakpoints[0]%
        if brightness_pct <= breakpoints[1]:
            return 1  # Stage 2: breakpoints[0]% to breakpoints[1]%
        if brightness_pct <= breakpoints[2]:
            return 2  # Stage 3: breakpoints[1]% to breakpoints[2]%
        return 3  # Stage 4: breakpoints[2]% to 100%

    def _calculate_zone_brightness_from_config(
        self,
        overall_pct: float,
        stage: int,
        zone_ranges: list[list[int]],
        breakpoints: list[int],
    ) -> float:
        """Calculate brightness for a zone based on configuration and current stage."""
        # Get the brightness range for this stage
        stage_range = zone_ranges[stage]

        # If the range is [0, 0], the zone should be off
        if stage_range[0] == 0 and stage_range[1] == 0:
            return 0.0

        # Calculate position within the current stage
        stage_boundaries = [
            (0, breakpoints[0]),  # Stage 1
            (breakpoints[0], breakpoints[1]),  # Stage 2
            (breakpoints[1], breakpoints[2]),  # Stage 3
            (breakpoints[2], 100),  # Stage 4
        ]

        stage_start, stage_end = stage_boundaries[stage]

        # Calculate progress within the stage (0.0 to 1.0)
        if stage_end == stage_start:
            progress = 0.0
        else:
            progress = max(
                0.0, min(1.0, (overall_pct - stage_start) / (stage_end - stage_start))
            )

        # Map progress to the configured brightness range for this zone
        min_brightness, max_brightness = stage_range

        # Apply brightness curve to the progress
        curve_type = get_config_value(
            self._entry, CONF_BRIGHTNESS_CURVE, DEFAULT_BRIGHTNESS_CURVE
        )
        curved_progress = self._apply_brightness_curve(progress, curve_type)

        return min_brightness + (curved_progress * (max_brightness - min_brightness))

    def _apply_brightness_curve(self, progress: float, curve_type: str) -> float:
        """Apply brightness curve to linear progress for more natural feel."""
        if curve_type == CURVE_QUADRATIC:
            # Gentle curve: more linear at very low values, curve kicks in higher
            # For very low progress (< 10%), use mostly linear
            if progress < 0.1:
                return progress * 0.9 + (progress**0.5) * 0.1
            # For higher progress, use more curve
            return 0.4 * progress + 0.6 * (progress**0.5)
        if curve_type == CURVE_CUBIC:
            # More aggressive curve for maximum low-end precision
            if progress < 0.1:
                return progress * 0.8 + (progress ** (1 / 3)) * 0.2
            return 0.2 * progress + 0.8 * (progress ** (1 / 3))
        # Linear curve: even response
        return progress

    async def _control_lights(
        self, light_entities: list[str], brightness_pct: float
    ) -> None:
        """Turn on lights with specified brightness."""
        brightness_value = int(brightness_pct * 255)

        for entity_id in light_entities:
            try:
                await self.hass.services.async_call(
                    "light",
                    "turn_on",
                    {
                        "entity_id": entity_id,
                        "brightness": brightness_value,
                    },
                )
            except (ServiceNotFound, ValueError) as err:
                _LOGGER.error("Failed to control light %s: %s", entity_id, err)

    async def _turn_off_lights(self, light_entities: list[str]) -> None:
        """Turn off lights."""
        for entity_id in light_entities:
            try:
                await self.hass.services.async_call(
                    "light",
                    "turn_off",
                    {"entity_id": entity_id},
                )
            except (ServiceNotFound, ValueError) as err:
                _LOGGER.error("Failed to turn off light %s: %s", entity_id, err)

    async def async_turn_off(self, **kwargs: Any) -> None:
        """Turn off the combined light."""
        self._attr_is_on = False

        # Turn off all configured lights
        all_lights = self._get_all_controlled_lights()

        if all_lights:
            await self._turn_off_lights(all_lights)

        _LOGGER.info("Combined light turned off - all configured lights turned off")

        self.async_write_ha_state()
