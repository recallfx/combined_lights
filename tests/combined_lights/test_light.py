"""Test the Combined Lights light platform."""

from unittest.mock import Mock

import pytest
from homeassistant.components.light import ColorMode
from homeassistant.const import CONF_NAME
from homeassistant.core import HomeAssistant

from custom_components.combined_lights.const import (
    CURVE_LINEAR,
    CURVE_QUADRATIC,
    DEFAULT_BREAKPOINTS,
    DEFAULT_STAGE_1_BRIGHTNESS_RANGES,
    DEFAULT_STAGE_2_BRIGHTNESS_RANGES,
    DEFAULT_STAGE_3_BRIGHTNESS_RANGES,
    DEFAULT_STAGE_4_BRIGHTNESS_RANGES,
)
from custom_components.combined_lights.light import (
    CombinedLight,
    async_setup_entry,
    get_brightness_ranges,
    get_config_value,
    get_light_zones,
)


class TestLightPlatformUtilities:
    """Test utility functions for the light platform."""

    def test_get_config_value(self, mock_config_entry_advanced) -> None:
        """Test getting configuration value with default fallback."""
        # Test existing value
        result = get_config_value(mock_config_entry_advanced, CONF_NAME, "Default Name")
        assert result == "Advanced Test Combined Lights"

        # Test default fallback
        result = get_config_value(
            mock_config_entry_advanced, "non_existent", "Default Value"
        )
        assert result == "Default Value"

    def test_get_light_zones(self, mock_config_entry_advanced) -> None:
        """Test getting light zones from configuration."""
        zones = get_light_zones(mock_config_entry_advanced)

        expected = {
            "stage_1": ["light.stage1_1", "light.stage1_2"],
            "stage_2": ["light.stage2_1"],
            "stage_3": [],
            "stage_4": ["light.stage4_1"],
        }
        assert zones == expected

    def test_get_brightness_ranges(self, mock_config_entry_advanced) -> None:
        """Test getting brightness ranges from configuration."""
        ranges = get_brightness_ranges(mock_config_entry_advanced)

        assert ranges["stage_1"] == [[10, 40], [50, 70], [80, 90], [95, 100]]
        assert ranges["stage_2"] == [[0, 0], [20, 50], [60, 80], [85, 100]]
        assert ranges["stage_3"] == [[0, 0], [0, 0], [30, 60], [70, 100]]
        assert ranges["stage_4"] == [[0, 0], [0, 0], [0, 0], [40, 100]]

    def test_get_brightness_ranges_with_defaults(self, mock_config_entry) -> None:
        """Test getting brightness ranges with default values."""
        ranges = get_brightness_ranges(mock_config_entry)

        assert ranges["stage_1"] == DEFAULT_STAGE_1_BRIGHTNESS_RANGES
        assert ranges["stage_2"] == DEFAULT_STAGE_2_BRIGHTNESS_RANGES
        assert ranges["stage_3"] == DEFAULT_STAGE_3_BRIGHTNESS_RANGES
        assert ranges["stage_4"] == DEFAULT_STAGE_4_BRIGHTNESS_RANGES


class TestAsyncSetupEntry:
    """Test async_setup_entry function."""

    async def test_async_setup_entry(
        self, hass: HomeAssistant, mock_config_entry
    ) -> None:
        """Test setting up the light platform."""
        mock_add_entities = Mock()

        await async_setup_entry(hass, mock_config_entry, mock_add_entities)

        # Verify that one entity was added
        mock_add_entities.assert_called_once()
        entities = mock_add_entities.call_args[0][0]
        assert len(entities) == 1
        assert isinstance(entities[0], CombinedLight)


class TestCombinedLight:
    """Test CombinedLight entity."""

    @pytest.fixture
    def combined_light(self, mock_config_entry_advanced) -> CombinedLight:
        """Create a CombinedLight instance for testing."""
        return CombinedLight(mock_config_entry_advanced)

    def test_init(
        self, combined_light: CombinedLight, mock_config_entry_advanced
    ) -> None:
        """Test CombinedLight initialization."""
        assert combined_light._attr_name == "Advanced Test Combined Lights"
        assert (
            combined_light._attr_unique_id
            == f"{mock_config_entry_advanced.entry_id}_combined_light"
        )
        assert combined_light._attr_is_on is False
        assert combined_light._attr_brightness == 255
        assert combined_light._attr_supported_color_modes == {ColorMode.BRIGHTNESS}
        assert combined_light._attr_color_mode == ColorMode.BRIGHTNESS

    def test_available(self, combined_light: CombinedLight) -> None:
        """Test entity availability."""
        assert combined_light.available is True

    def test_get_stage_from_brightness(self, combined_light: CombinedLight) -> None:
        """Test determining stage from brightness percentage."""

        # Test each stage using the brightness calculator
        brightness_calc = combined_light._brightness_calc
        assert brightness_calc.get_stage_from_brightness(10) == 0  # Stage 1
        assert brightness_calc.get_stage_from_brightness(30) == 0  # Stage 1
        assert brightness_calc.get_stage_from_brightness(35) == 1  # Stage 2
        assert brightness_calc.get_stage_from_brightness(60) == 1  # Stage 2
        assert brightness_calc.get_stage_from_brightness(65) == 2  # Stage 3
        assert brightness_calc.get_stage_from_brightness(90) == 2  # Stage 3
        assert brightness_calc.get_stage_from_brightness(95) == 3  # Stage 4
        assert brightness_calc.get_stage_from_brightness(100) == 3  # Stage 4

    def test_apply_brightness_curve_linear(self, combined_light: CombinedLight) -> None:
        """Test linear brightness curve."""
        brightness_calc = combined_light._brightness_calc
        # Linear curve should return input unchanged
        assert brightness_calc._apply_brightness_curve(0.0, CURVE_LINEAR) == 0.0
        assert brightness_calc._apply_brightness_curve(0.5, CURVE_LINEAR) == 0.5
        assert brightness_calc._apply_brightness_curve(1.0, CURVE_LINEAR) == 1.0

    def test_apply_brightness_curve_quadratic(
        self, combined_light: CombinedLight
    ) -> None:
        """Test quadratic brightness curve."""
        brightness_calc = combined_light._brightness_calc
        # Quadratic curve should be different from linear for mid-range values
        result = brightness_calc._apply_brightness_curve(0.5, CURVE_QUADRATIC)
        assert result != 0.5  # Should be different from linear
        assert 0.0 <= result <= 1.0  # Should be within valid range

    def test_calculate_zone_brightness_off_range(
        self, combined_light: CombinedLight
    ) -> None:
        """Test zone brightness calculation for off range [0, 0]."""
        brightness_calc = combined_light._brightness_calc
        # Zone should be off when range is [0, 0]
        result = brightness_calc._calculate_zone_brightness_from_config(
            50.0,  # overall_pct
            1,  # stage
            [[0, 0], [0, 0], [0, 0], [0, 0]],  # zone_ranges (all off)
            DEFAULT_BREAKPOINTS,
        )
        assert result == 0.0

    def test_calculate_zone_brightness_active_range(
        self, combined_light: CombinedLight
    ) -> None:
        """Test zone brightness calculation for active range."""
        brightness_calc = combined_light._brightness_calc
        # Zone should have brightness when range is not [0, 0]
        result = brightness_calc._calculate_zone_brightness_from_config(
            50.0,  # overall_pct (exactly at breakpoint[1])
            1,  # stage 2
            [[0, 0], [10, 50], [60, 80], [90, 100]],  # zone_ranges
            DEFAULT_BREAKPOINTS,  # [25, 50, 75]
        )
        # Should be maximum of stage 2 range since we're at the upper boundary
        assert result == 50.0

    def test_get_all_controlled_lights(self, combined_light: CombinedLight) -> None:
        """Test getting all controlled light entity IDs."""
        zone_manager = combined_light._zone_manager
        lights = zone_manager.get_all_lights()
        expected = [
            "light.stage1_1",
            "light.stage1_2",
            "light.stage2_1",
            "light.stage4_1",
        ]
        assert sorted(lights) == sorted(expected)

    def test_get_average_brightness_mocked(self, combined_light: CombinedLight) -> None:
        """Test getting average brightness with mocked states."""
        # Mock the hass object and states
        combined_light.hass = Mock()

        def mock_get_state(entity_id):
            if entity_id == "light.test1":
                return Mock(state="on", attributes={"brightness": 100})
            elif entity_id == "light.test2":
                return Mock(state="on", attributes={"brightness": 200})
            return Mock(state="off")

        combined_light.hass.states.get = mock_get_state

        zone_manager = combined_light._zone_manager
        result = zone_manager.get_average_brightness(
            combined_light.hass, ["light.test1", "light.test2", "light.test3"]
        )
        assert result == 150  # (100 + 200) / 2

    def test_get_average_brightness_all_off_mocked(
        self, combined_light: CombinedLight
    ) -> None:
        """Test getting average brightness when all lights are off."""
        # Mock the hass object and states
        combined_light.hass = Mock()
        combined_light.hass.states.get = Mock(return_value=Mock(state="off"))

        zone_manager = combined_light._zone_manager
        result = zone_manager.get_average_brightness(
            combined_light.hass, ["light.test1", "light.test2"]
        )
        assert result is None

    def test_is_on_property_mocked(self, combined_light: CombinedLight) -> None:
        """Test is_on property with mocked states."""
        # Mock the hass object and states
        combined_light.hass = Mock()

        # Test when all lights are off
        combined_light.hass.states.get = Mock(return_value=Mock(state="off"))
        assert combined_light.is_on is False

        # Test when some lights are on
        def mock_get_state(entity_id):
            if entity_id == "light.stage1_1":
                return Mock(state="on")
            return Mock(state="off")

        combined_light.hass.states.get = mock_get_state
        assert combined_light.is_on is True

    def test_brightness_property(self, combined_light: CombinedLight) -> None:
        """Test brightness property."""
        # Mock is_on to control the property behavior
        combined_light.hass = Mock()
        combined_light.hass.states.get = Mock(return_value=Mock(state="off"))

        # When off, brightness should be None
        assert combined_light.brightness is None

        # When on, should return target brightness
        combined_light.hass.states.get = Mock(return_value=Mock(state="on"))
        combined_light._target_brightness = 128
        assert combined_light.brightness == 128

    def test_entity_state_snapshot(
        self, combined_light: CombinedLight, snapshot
    ) -> None:
        """Test entity state attributes with snapshot testing."""
        # Set up some specific state
        combined_light._target_brightness = 150
        combined_light._attr_is_on = True

        # Test that the entity attributes match snapshot (excluding dynamic unique_id)
        entity_attributes = {
            "name": combined_light._attr_name,
            "is_on": combined_light._attr_is_on,
            "brightness": combined_light._attr_brightness,
            "supported_color_modes": list(combined_light._attr_supported_color_modes),
            "color_mode": combined_light._attr_color_mode,
            "target_brightness": combined_light._target_brightness,
        }

        assert entity_attributes == snapshot
