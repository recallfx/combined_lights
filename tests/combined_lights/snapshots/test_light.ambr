# serializer version: 1
# name: TestCombinedLight.test_entity_state_snapshot
  dict({
    'brightness': 255,
    'color_mode': <ColorMode.BRIGHTNESS: 'brightness'>,
    'is_on': True,
    'name': 'Advanced Test Combined Lights',
    'supported_color_modes': list([
      <ColorMode.BRIGHTNESS: 'brightness'>,
    ]),
    'target_brightness': 150,
  })
# ---
